staging/src/k8s.io/apimachinery/pkg/util/net/interface.go

```diff
@@ -149,90 +163,106 @@ func getIPFromInterface(intfName string, nw networkInterfacer) (net.IP, error) {
 	return nil, nil
 }
 
-func flagsSet(flags net.Flags, test net.Flags) bool {
-	return flags&test != 0
-}
-
-func flagsClear(flags net.Flags, test net.Flags) bool {
-	return flags&test == 0
+// memberOF tells if the IP is of the desired family. Used for checking interface addresses.
+func memberOf(ip net.IP, family AddressFamily) bool {
+	ipFamily := familyIPv4
+	if ip.To4() == nil {
+		ipFamily = familyIPv6
+	}
+	return ipFamily == family
 }
 
-func chooseHostInterfaceNativeGo() (net.IP, error) {
-	intfs, err := net.Interfaces()
+// chooseIPFromHostInterfaces looks at all system interfaces, trying to find one that is up that
+// has a global IP, and returns the IP. Searches for IPv4 addresses, and then IPv6 addresses.
+func chooseIPFromHostInterfaces(nw networkInterfacer) (net.IP, error) {
+	intfs, err := nw.Interfaces()
 	if err != nil {
 		return nil, err
 	}
-	i := 0
-	var ip net.IP
-	for i = range intfs {
-		if flagsSet(intfs[i].Flags, net.FlagUp) && flagsClear(intfs[i].Flags, net.FlagLoopback|net.FlagPointToPoint) {
-			addrs, err := intfs[i].Addrs()
+	if len(intfs) == 0 {
+		return nil, fmt.Errorf("no interfaces found on host.")
+	}
+	skipReason := ""
+	for _, family := range []AddressFamily{familyIPv4, familyIPv6} {
+		glog.V(4).Infof("Looking for system interface with a global IPv%d address", uint(family))
+		for _, intf := range intfs {
+			if !isInterfaceUp(&intf) {
+				skipReason = "down interface"
+				glog.V(4).Infof("Skipping: %s %q", skipReason, intf.Name)
+				continue
+			}
+			if isLoopbackOrPointToPoint(&intf) {
+				skipReason = "LB or P2P interface"
+				glog.V(4).Infof("Skipping: %s %q", skipReason, intf.Name)
+				continue
+			}
+			cidrs, err := nw.Addrs(&intf)
 			if err != nil {
 				return nil, err
 			}
-			if len(addrs) > 0 {
-				for _, addr := range addrs {
-					if addrIP, _, err := net.ParseCIDR(addr.String()); err == nil {
-						if addrIP.To4() != nil {
-							ip = addrIP.To4()
-							if !ip.IsLinkLocalMulticast() && !ip.IsLinkLocalUnicast() {
-								break
-							}
-						}
-					}
+			if len(cidrs) == 0 {
+				skipReason = "no addresses on interface"
+				glog.V(4).Infof("Skipping: %s %q", skipReason, intf.Name)
+				continue
+			}
+			for _, cidr := range cidrs {
+				ip, _, err := net.ParseCIDR(cidr.String())
```

Comment:
OK. So I guess the issue is with the naming, which is very vague (CIDR vs address, vs IP, vs NetIP). From what I can tell, nw.Addrs() returns a *netlink.Addr* struct which has a NetIP field and a String() method that returns the IP/netmask (what I call a CIDR). It doesn't return a net.Addr type from what I can tell.

I can change it back to addr and addrs to avoid (maybe) confusion.